jobhub.go

```diff
@@ -122,106 +109,95 @@ func (p *Pipeline) AddJob(job Job) Job {
 	job.id = p.nextIDJob()
 	p.jobContainer = append(p.jobContainer, job)
 	p.jobByID[job.id] = job
-	p.startingJob[job.id] = true
-	return job
+	return job, nil
 }
 
-func (p *Pipeline) AddJobDependency(job Job, deps ...Job) {
-	var tempContainer []Job
-	tempContainer = append(deps, job)
+func (p *pipeline) AddJobDependency(job Job, deps ...Job) error {
+	tempContainer := append(deps, job)
 	if p.id == 0 {
-		p.Log.Panicf("Pipeline [%d][%s] | Pipeline not initialized",
-			p.id, p.Name)
-	}
-	if p.jobContainer == nil {
-		p.Log.Panicf("Pipeline [%d][%s] | Job Container is empty",
-			p.id, p.Name)
+		return fmt.Errorf("%s (%d) not initialized (i.e. has no jobs added)", p.name, p.id)
 	}
-	for _, givenJob := range tempContainer {
-		if givenJob.pipelineID != p.id {
-			p.Log.Panicf("Pipeline [%d][%s] | Job [%d][%s] does not belong to this pipeline",
-				p.id, p.Name, givenJob.id, givenJob.Name)
+	for _, j := range tempContainer {
+		if j.pipelineID != p.id {
+			return fmt.Errorf("%s (%d) does not belong to %s (%d)", j.Name, j.id, p.name, p.id)
 		}
 	}
 	for _, d := range deps {
 		p.jobDependency[job.id] = append(p.jobDependency[job.id], d.id)
-		delete(p.startingJob, d.id)
 	}
+	return nil
 }
 
-func (p *Pipeline) resolveDependencyRecursion(jobID, level int) {
-	if l := p.recursionLevels[jobID]; l >= level {
-		return
-	}
-	p.recursionLevels[jobID] = level
-	level++
-	for _, depID := range p.jobDependency[jobID] {
-		p.resolveDependencyRecursion(depID, level)
-	}
-}
+func (p pipeline) topologicalSort() ([]int, error) {
+	var (
+		temporaryMark = map[int]bool{}
```

Comment:
What about we use `make` for all new slices we don't plant to have initial values? 
This adds info about the intention "I want an empty map".

Reply:
Done.